posts/2022-10-04-microprofile-serverless-ibm-code-engine.adoc

```diff
@@ -0,0 +1,414 @@
+
+When run within the context of Open Liberty, the `CloudEventsProvider` class is automatically configured for marshalling and unmarshalling of CloudEvents.
+
+With the libary imported the application can now be updated to make use of CloudEvents.
+
+You can view the <<completed_class>> to check that changes made through the following sections align with expected result.
+
+==== Returning a CloudEvent
+Having added the necessary library, how do we firstly return a CloudEvent.
+
+If the Getting Started guide was used, the SystemResource class is a good candidate to convert the response Type from `Responce` to `CloudEvent`
+
+For the Method Declaration replace `Response` with `CloudEvent`
+```
+public CloudEvent getProperties() {
+```
+
+Having changed the output type we now need to construct a CloudEvent to return. However we need to do some work on the system properties to be able to include them as the data within the event.
+
+CloudEvents requires the data to be in a binary format when it is provided during its building process as it itself cannot do the conversion from Object to byte[]. As such we can take the properties from `System.getProperties()` and make them into a JSON string with use of Jsonb.
+```
+/* java.util.properties does not have a direct way to obtain a byte[] so store in an intermediary Map first*/
+    Map properties = System.getProperties();
+    Jsonb jsonb = JsonbBuilder.create();
+/* convert properties map into a JSON string which can then be converted into a byte[]*/
+    String jsonString = jsonb.toJson(properties);
+
+```
+
+With our Data in String format, we can now get the byte[] representation of the data.
+
+The `CloudEventBuilder` class provides the necessary components to build our CloudEvent and for this, it is recommended you use `v1()` as that is the most recent specification version.
+```
+    return CloudEventBuilder.v1()
+        .withData(jsonString.getBytes())
+        .withDataContentType("application/json")
+        .withId("properties")
+        .withType("java.properties")
+        .withSource(URI.create("http://system.poperties"))
+        .build();
+```
+Aside from `withData()`, the reset of the methods set the values that will be returned as headers in the response.
+Once all required properties are set you can build the CloudEvent Object.
+
+If you are missing any required properties, when your application attempts to build the CloudEvent an exception will be thrown informing of a missing property. To see which properties are required you can review the https://github.com/cloudevents/spec/blob/v1.0/spec.md[specification]
+
+As CloudEvents can be from a wide variety of sources can that differ even within a single provider the majority of fields are fairly free-form.
```

Comment:
```suggestion
Because CloudEvents can be from a wide variety of sources  that might differ even within a single provider, the majority of the fields are fairly free-form.
```